flink-ml-core/src/main/java/org/apache/flink/ml/util/QuantileSummary.java

```diff
@@ -0,0 +1,359 @@
+ * Greenwald, Michael and Khanna, Sanjeev. (https://doi.org/10.1145/375663.375670)
+ */
+public class QuantileSummary {
+
+    private static final int BUFFER_SIZE = 10000;
+    private static final int COMPRESS_THRESHOLD = 10000;
+
+    private double relativeError;
+    private long count;
+    private List<StatsTuple> sampled;
+    private List<Double> headBuffer = new ArrayList<>(BUFFER_SIZE);
+
+    /**
+     * * QuantileSummaries Constructor.
+     *
+     * @param relativeError The target relative error.
+     */
+    public QuantileSummary(double relativeError) {
+        this(relativeError, Collections.EMPTY_LIST, 0);
+    }
+
+    /**
+     * QuantileSummaries Constructor.
+     *
+     * @param relativeError The target relative error.
+     * @param sampled A buffer of quantile statistics. See the G-K article for more details.
+     * @param count The count of all the elements inserted in the sampled buffer.
+     */
+    public QuantileSummary(double relativeError, List<StatsTuple> sampled, long count) {
+        if (relativeError <= 0 || relativeError >= 1) {
+            throw new RuntimeException("An appropriate relative error must lay between 0 and 1.");
+        }
+        this.relativeError = relativeError;
+        this.sampled = sampled;
+        this.count = count;
+    }
+
+    /**
+     * Insert a new observation to the summary.
+     *
+     * @param item The new observation to insert into the summary.
+     * @return A summary with the given observation inserted into the summary.
+     */
+    public QuantileSummary insert(Double item) {
+        headBuffer.add(item);
+        if (headBuffer.size() >= BUFFER_SIZE) {
+            QuantileSummary result = insertHeadBuffer();
+            if (result.sampled.size() >= COMPRESS_THRESHOLD) {
+                return result.compress();
+            } else {
+                return result;
+            }
+        } else {
+            return this;
+        }
+    }
+
+    /**
+     * This implements the COMPRESS function of the GK algorithm.
+     *
+     * @return A new summary that compresses the summary statistics and the head buffer.
+     */
+    public QuantileSummary compress() {
+        QuantileSummary inserted = insertHeadBuffer();
+        Preconditions.checkState(inserted.headBuffer.isEmpty());
+        Preconditions.checkState(inserted.count == count + headBuffer.size());
+
+        List<StatsTuple> compressed =
+                compressInternal(inserted.sampled, 2 * relativeError * inserted.count);
+        return new QuantileSummary(relativeError, compressed, inserted.count);
+    }
+
+    /**
+     * * Merges two compressed summaries together.
+     *
+     * @param other The summary to be merged.
+     * @return The merged summary.
+     */
+    public QuantileSummary merge(QuantileSummary other) {
+        checkState(headBuffer.isEmpty(), "Current buffer needs to be compressed before merge");
```

Comment:
nit: The usage of `checkState` and `Preconditions.checkState` seems inconsistent in this PR.